File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/mutation_client.R
#' @title Mutation Client
#' @description Mutation client for executing write operations
#' @name MutationClient
NULL

#' @importFrom R6 R6Class
#' @export
MutationClient <- R6::R6Class(
  "MutationClient",
  
  public = list(
    #' @description Create a new mutation client
    #' @param http_client HttpClient instance
    initialize = function(http_client) {
      private$http <- http_client
    },
    
    #' @description Execute a mutation operation
    #' @param mutation_name The name of the mutation
    #' @param input Mutation input data
    #' @return grey_result with mutation result or error
    mutate = function(mutation_name, input) {
      if (is.null(mutation_name) || nchar(trimws(mutation_name)) == 0) {
        return(result_err(error_validation("Mutation name is required")))
      }
      
      if (is.null(input)) {
        return(result_err(error_validation("Input data is required")))
      }
      
      body <- list(
        mutation = mutation_name,
        input = input
      )
      
      private$http$post("/api/v1/mutate", body)
    },
    
    #' @description Execute a batch of mutations
    #' @param mutations List of mutation objects with mutation_name and input
    #' @return grey_result with array of results or error
    batch_mutate = function(mutations) {
      if (is.null(mutations) || !is.list(mutations)) {
        return(result_err(error_validation("Mutations list is required")))
      }
      
      if (length(mutations) == 0) {
        return(result_err(error_validation("At least one mutation is required")))
      }
      
      formatted <- lapply(mutations, function(m) {
        list(
          mutation = m$mutation_name,
          input = m$input
        )
      })
      
      private$http$post("/api/v1/mutate/batch", list(mutations = formatted))
    },
    
    #' @description Execute a mutation within a transaction
    #' @param transaction_id Transaction ID
    #' @param mutation_name Mutation name
    #' @param input Mutation input
    #' @return grey_result with mutation result or error
    transactional_mutate = function(transaction_id, mutation_name, input) {
      if (is.null(transaction_id) || nchar(trimws(transaction_id)) == 0) {
        return(result_err(error_validation("Transaction ID is required")))
      }
      
      if (is.null(mutation_name) || nchar(trimws(mutation_name)) == 0) {
        return(result_err(error_validation("Mutation name is required")))
      }
      
      if (is.null(input)) {
        return(result_err(error_validation("Input data is required")))
      }
      
      body <- list(
        transaction_id = transaction_id,
        mutation = mutation_name,
        input = input
      )
      
      private$http$post("/api/v1/mutate/transactional", body)
    }
  ),
  
  private = list(
    http = NULL
  )
)

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/domain/mutation_client.rb
# frozen_string_literal: true

module GreySdk
  module Domain
    # Mutation client for executing write operations
    class MutationClient
      # @param http_client [GreySdk::Http::HttpClient] The HTTP client
      def initialize(http_client)
        @http = http_client
      end

      # Executes a mutation operation
      # @param mutation_name [String] The name of the mutation to execute
      # @param input [Hash] Mutation input data
      # @return [GreySdk::Error::Result] Result containing mutation result or error
      def mutate(mutation_name:, input:)
        return validation_error("Mutation name is required") if mutation_name.nil? || mutation_name.empty?
        return validation_error("Input data is required") if input.nil?

        body = {
          mutation: mutation_name,
          input: input
        }

        @http.post("/api/v1/mutate", body)
      end

      # Executes a batch of mutations
      # @param mutations [Array<Hash>] Array of mutation objects with :mutation_name and :input
      # @return [GreySdk::Error::Result] Result containing array of results or error
      def batch_mutate(mutations:)
        return validation_error("Mutations array is required") if mutations.nil? || !mutations.is_a?(Array)
        return validation_error("At least one mutation is required") if mutations.empty?

        body = {
          mutations: mutations.map do |m|
            {
              mutation: m[:mutation_name],
              input: m[:input]
            }
          end
        }

        @http.post("/api/v1/mutate/batch", body)
      end

      # Executes a mutation within a transaction
      # @param transaction_id [String] Transaction ID
      # @param mutation_name [String] Mutation name
      # @param input [Hash] Mutation input
      # @return [GreySdk::Error::Result] Result containing mutation result or error
      def transactional_mutate(transaction_id:, mutation_name:, input:)
        return validation_error("Transaction ID is required") if transaction_id.nil? || transaction_id.empty?
        return validation_error("Mutation name is required") if mutation_name.nil? || mutation_name.empty?
        return validation_error("Input data is required") if input.nil?

        body = {
          transaction_id: transaction_id,
          mutation: mutation_name,
          input: input
        }

        @http.post("/api/v1/mutate/transactional", body)
      end

      private

      def validation_error(message)
        Error::Result.err(Error::GreyError.validation(message))
      end
    end
  end
end
